### ISSUE TITLE
nha.mod missing in CMake builds

### ISSUE THREAD

**Issue by MathiasMagnus** [Amber-MD/cmake-buildscripts#131]
I'm trying to build Amber using CMake and putting aside my thoughts of the overall architecture (I'll try to devise a way to improve CMake build and make it far more robust and intuitive). I had to tweak the build scripts on a multitude of places but this is the error I get when I try to build:
```
/home/mnagy/Source/Repos/amber18/AmberTools/src/pbsa/irre31.F90:19:22:                                                                                                                                     
          lwar=3*inmax*inmax/2+10*inmax+2*immax+1, &
                      1                                                                                                                                                                                    Warning: Integer division truncated to constant ‘1093’ at (1) [-Winteger-division]                                                                                                                         /home/mnagy/Source/Repos/amber18/AmberTools/src/pbsa/pb_fdfrc.F90:21:7:                                                                                                                                    
    use nha, only : pb_nhadrv
       1                                                                                                                                                                                                   Fatal Error: Can't open module file ‘nha.mod’ for reading at (1): No such file or directory
```
How is `nha.mod` supposed to manifest? I don't see any targets that supposed to produce this byproduct.

**Comment by multiplemonomials** [comment]
Hey, are you interested in helping out with CMake development?  We could use the help.  Hang on I'm taking a look.

**Comment by multiplemonomials** [comment]
Huh as far as I can tell, `irre31.F90` doesn't refer to `nha`.  Is it possible that you are doing a parallel build and multiple compiler warnings/errors got squished together?  Try again doing a single build (`make -j1`)

**Comment by MathiasMagnus** [comment]
Indeed, when using CMake I was using the Ninja generator which builds in parallel by default. Because I couldn't get the CMake build working I fell back to autotools and parallel builds don't work there either. The subproject builds are hardcoded to be serial, and indeed, trying anything other than -j1 breaks the build.

Both the CMake and autotools builds have implicit dependencies, meaning not all dependencies are marked explicitly in the makefiles, hence the order of build matters. Both the autotools and CMake scripts would require a facelift.

I'm definitely in favor of fixing the CMake scripts and the source where needed to get faster and more reliable builds, plus hook code navigation tooling into it. Autotools lack compile_commands.json support which is required for code navigation tools. If I get the green lights, I'll enroll the scripts into my "pimp yo CMake" program.

My general impression of both build systems are that they do too much. Dependency detection and on-demand installation is flaky. I'd highly suggest taking it out as it is and delegate building dependencies to projects meant to do just that. Vcpkg or Conan. I have more experience with the prior and it helps a lot.

```
vcpkg install boost-filesystem fftw3 any-other-sexy-dep-from-a-1000-other-packages
cmake -D CMAKE_TOOLCHAIN_FILE=$(VCPKG_ROOT)/scripts/buildsystem/vcpkg.cmake ~/amber18
cmake --build .
```

Something like this. The scripts need not be made specific to the package manager, it is vanilla CMake, the toolchain file machinery is (ab)used to set extra `CMAKE_PREFIX_PATH`s and what not. It works very well and much more reliably than anything you can cook inside a project. (I would argue that sooner or later the self-hosted dependencies might be made obsolete as well.) Package managers would also help in getting native Windows builds with MSVC or Clang and not through MinGW.

**Comment by multiplemonomials** [comment]
Huh using a package manager does sound interesting, though in my experience it can be quite difficult to convince Amber developers to switch to new frameworks like that -- just look at how many years it's taken me to get us close to adopting CMake.  Also, it would need to provide similar features and supported platforms.  Do you know if these package managers are supported/installable on the more esoteric high performance computing environments that Amber runs on, such as Cray supercomputers?  Also, can they work with alternate compilers such as Intel and PGI?

Anyway, can you try again using `ninja -j1` and see if you can get the full error message?

Also, what's this compile_commands.json thing?

**Comment by MathiasMagnus** [comment]
I'll look into the esoteric platform support. It shouldn't cause much of an issue but I'll make sure.

As for compile_commands.json is nothing more than a JSON file which holds ALL of the compiler invocations with all of the compiler flags. Tools use it to light up code navigation and IntelliSense inside IDEs. Right click a function, go to definition, go to declaration, expand macros, etc. It is what's required to make 100% accurate semantic representation of C/C++. Mighty useful for exploring code, debugging, get compiler errors without actually compiling the code... Makes development much faster.

There are other tools available, compile_commands.json is just one of them. CMake Server Mode (now deprecated) and the newer [CMake File API](https://cmake.org/cmake/help/v3.15/manual/cmake-file-api.7.html#manual:cmake-file-api(7)) may be used for this as well. The IDE can "ask" CMake questions and CMake replies.

**Comment by multiplemonomials** [comment]
OK so there is a bit of a story behind these issues (saw your email, it reminded be that this needed fixing).  A month or two before AmberTools 19 released, I heard from other developers that the CMake build was broken.  It turned out that unbeknownst to me, people had turned off my CMake CI builds, so these issues were not detected for months and a substantial amount of differences between the source code and the CMake build system had built up.  I stared working on a PR to fix this, but unfortunately I was not fast enough and was not able to get it merged before the code freeze, and the CMake build being broken was not considered a serious enough issue to be worth merging into frozen code.  So that's how we got the current situation where AmberTools 19 has a nonfunctional CMake build system -- I was not fast enough with merging the needed changes.  

There is something I can do now, and I'm such and idiot that I didn't think of this before.  I generated a patch based on the PR I merged to fix all of the CMake issues -- you should be able to merge this to AmberTools 19 and that should get the CMake build system in working order.  There might be a few areas that don't apply due to areas that are not included in the public release build, but you can safely ignore those.  

@MathiasMagnus give this patch a try and let me know if it works: [amber19-fix-cmake.patch.zip](https://github.com/Amber-MD/cmake-buildscripts/files/3917907/amber19-fix-cmake.patch.zip)
